zkvm/src/signature/prover.rs

```diff
@@ -0,0 +1,191 @@
+            },
+            self.R_i,
+        )
+    }
+}
+
+impl PartyAwaitingCommitments {
+    pub fn receive_commitments(
+        mut self,
+        nonce_commitments: Vec<NonceCommitment>,
+    ) -> Result<(PartyAwaitingSiglets, Siglet), VMError> {
+        // Check stored precommitments against received commitments
+        for (pre_comm, comm) in self
+            .nonce_precommitments
+            .iter()
+            .zip(nonce_commitments.iter())
+        {
+            // Make H(comm) = H(R_i)
+            let actual_precomm = comm.precommit();
+
+            // Compare H(comm) with pre_comm, they should be equal
+            // TBD: should we use ct_eq?
+            let equal = pre_comm.0.ct_eq(&actual_precomm.0);
+            if equal.unwrap_u8() == 0 {
+                return Err(VMError::InconsistentWitness);
+            }
+        }
+
+        // Make R = sum_i(R_i). nonce_commitments = R_i from all the parties.
+        let R: RistrettoPoint = nonce_commitments.iter().map(|R_i| R_i.0).sum();
+
+        // Make c = H(X_agg, R, m)
+        // The message should already have been fed into the transcript.
+        let c = {
+            self.transcript
+                .commit_point(b"X_agg", &self.multikey.aggregated_key().0);
+            self.transcript.commit_point(b"R", &R.compress());
+            self.transcript.challenge_scalar(b"c")
+        };
+
+        // Make a_i = H(L, X_i)
+        let X_i = VerificationKey((self.x_i.0 * RISTRETTO_BASEPOINT_POINT).compress());
+        let a_i = self.multikey.factor_for_key(&X_i);
+
+        // Generate siglet: s_i = r_i + c * a_i * x_i
+        let s_i = self.r_i.0 + c * a_i * self.x_i.0;
+
+        // Store received nonce commitments in next state
+        Ok((
+            PartyAwaitingSiglets {
+                multikey: self.multikey,
+                nonce_commitments,
+                c,
+            },
+            Siglet(s_i),
+        ))
+    }
+}
+
+impl PartyAwaitingSiglets {
+    pub fn receive_siglets(self, siglets: Vec<Siglet>) -> Signature {
+        // s = sum(siglets)
+        let s: Scalar = siglets.iter().map(|siglet| siglet.0).sum();
+        // R = sum(R_i). nonce_commitments = R_i
+        let R: RistrettoPoint = self.nonce_commitments.iter().map(|R_i| R_i.0).sum();
+
+        Signature { s, R }
+    }
+
+    pub fn receive_and_verify_siglets(
+        self,
+        siglets: Vec<Siglet>,
+        pubkeys: Vec<PubKey>,
+    ) -> Signature {
+        // Check that all siglets are valid
+        for (i, s_i) in siglets.iter().enumerate() {
+            let S_i = s_i.0 * RISTRETTO_BASEPOINT_POINT;
+            let X_i = pubkeys[i].0;
+            let R_i = self.nonce_commitments[i].0;
+
+            // Make a_i = H(L, X_i)
+            let a_i = self
+                .multikey
+                .factor_for_key(&VerificationKey(X_i.compress()));
+
+            // Check that S_i = R_i + c * a_i * X_i
+            assert_eq!(S_i, R_i + self.c * a_i * X_i);
```

Comment:
let's fail with Result<> instead of panic?

Reply:
the error should also return the index of the party that failed

Reply:
sounds good!